<a href="https://colab.research.google.com/github/Struth-Rourke/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Lecture_Statistics_Probability_and_Inference.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>

<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Lesson 1*

# Statistics, Probability and Inference

## Learning Objectives
* [Part 1](#p1): Normal Distribution Revisted
* [Part 2](#p2): Student's T Test
* [Part 3](#p3): Hypothesis Test & Doing it Live

## What is Descriptive Statistics?

<https://statistics.laerd.com/statistical-guides/descriptive-inferential-statistics.php>

In [2]:
import pandas as pd
df = pd.DataFrame({'a': [1,2,3,4,5], 'b': [2,4,6,8,10]})
df.head()

a   b
0  1   2
1  2   4
2  3   6
3  4   8
4  5  10

In [3]:
# How can we quickly look at some descriptive statistics of the above dataframe?

# Descriptive stats describe a variables: center (median, mean); spread (variance, std dev); distribution (ranges, quartiles, etc.)
df.describe()

a          b
count  5.000000   5.000000
mean   3.000000   6.000000
std    1.581139   3.162278
min    1.000000   2.000000
25%    2.000000   4.000000
50%    3.000000   6.000000
75%    4.000000   8.000000
max    5.000000  10.000000

## What is Inferential Statistics?

<https://slideplayer.com/slide/5130463/16/images/2/Statistical+Inference.jpg>

## Hypothesis Testing (T-Tests)

Drawing conclusions based on probability in the face of randomness



Is the sample reflective of our concept of the population?

In [13]:
# Coinflips example - What is the population?
import numpy as np

coinflips = np.random.binomial(n = 1, p = .5, size = 10)

coinflips.mean()

0.506

In [14]:
pd.Series(coinflips).describe()

count    1000.000000
mean        0.506000
std         0.500214
min         0.000000
25%         0.000000
50%         1.000000
75%         1.000000
max         1.000000
dtype: float64

If we repeat the sampling why does our estimate of the mean move around so much? 

In [15]:
# Coinflips example - See the implications of randomness

coinflips = np.random.binomial(n = 1, p = .5, size = 1000)

coinflips.mean()

0.492

What happens as I increase the sample size? The sample mean deviates less from the...?

In [0]:
# More Information -> Greater Confidence in the face of randomness

coinflips = np.random.binomial(n = 1, p = .5, size = 1000)

# Sample Mean
coinflips.mean()

## Standard Error 

Quantifies how much the sample means moves

s = sample standard dev
n = our sample size

As sample size increases, the SE goes down; vice verse, since amount greatly affects SE of the overall sample

![Standard Error of the Sample  Mean](https://jungminded.weebly.com/uploads/1/3/5/8/13582285/7285766_orig.jpeg?160)

**Conclusion: As our sample size goes up, our estimates (of the mean in this case) get more precise, allowing us to infer things about the population with greater confidence.**



## One Sample T-test

What if I didn't have a fair coin?

Comparing information from a sample against our concept of the population.

![One Sample T Statistic](https://lh3.googleusercontent.com/proxy/CXO8J4hOx3DqUt-vfRfmKN-B9IL0wNZlnbQ6lXnopiAAGQWtLIY-f4qD_2zCX_C3p_aaqi9-Qni2MqUD8mVbnnXWEy-AtAywlEp7bUZuMWIZwmvbFJFhdeHcUY5v-i3vcdg5cL9jdh6votwdHnFd44uE)

## Let's Play a Game

Head = Give Gwen a dollar; 
Tails = Gwen gives me a dollar

## The T-test Process:

T- Test: Compares a sample to a concept of a population, and tells me whether or not the sample is reflective of the population or not.

Concept of the Population = Fair coin; p = .5

In [20]:
## Let's Play a Game

# 1 = Heads 
coinflips = np.random.binomial(n = 1, p = .6, size = 10)
coinflips 


array([1, 0, 1, 1, 1, 1, 0, 1, 1, 1])

## Establish the parts of a 1-sample T-test

**Null Hypothesis:** This is a fair coin. My concept of the population is correct $p(heads) = .5$

$H_0: \mu = \bar{x}$

My population mean ($\mu$) is equal to my sample mean ($\bar{x}$)

**Alternative Hypothesis:** (The opposite) My population mean is different than my sample mean

$H_a: \mu \neq \bar{x}$

**Confidence Level:** 95% Confidence Level = I am willing to be tricked by the randomness of the process 5% of the time; we could draw the wrong conclusion 5% of the time due to the randomness of the sampling. 

I want to be 95% confident that my null is false before I reject it; opposite is true as well, I want to be less than 5% confident that my null is true, before I will reject it

In [22]:
from scipy import stats

# Sample, Null Hypothesis Value (concept of the population, fair coin = .5p of heads)
# Is my sample a reflection of my concept of the population
# If there's a low probability that my sample is representative of the population -- then reject the null that p = .5
stats.ttest_1samp(coinflips, .5)

Ttest_1sampResult(statistic=2.25, pvalue=0.051003260706950763)

**T-stat:** 2.25

**P-value:** .0510

Probability that the Null is true, given the sample that I have collected/observed

5.1%

Based on this p-value, should I reject or fail to reject the null?

*Critical Value* = Probability threshold for reject vs fail to reject 
(1 - Confidence Interval) i.e (1 - .95 = .05)

p-value >= .05 --> Fail to reject

p-value < .05 --> Reject the null

**Conclusion:** Based on a p-value of .051, I fail to reject the null that my sample mean of coinflips is equal to my population mean of coinflips. I cannot say that this isn't a fair coin.

In [51]:
# Increase Sample Size

coinflips = np.random.binomial(n = 1, p = .6, size = 1000)

stats.ttest_1samp(coinflips, .5) 

Ttest_1sampResult(statistic=5.188868747067806, pvalue=2.5618587469830977e-07)

**Conclusion:** Based on a p-value of .00000025, I reject the null that my sample mean of coinflips is equal to my population mean of coinflips. I can say that this isn't a fair coin. Supports the alternative hypothesis.

In [55]:
coinflips[:100]

array([1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0,
       1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1])

## As our sample mean gets further away from the null (population mean), the p-value will decrease

In [60]:
coinflips = np.random.binomial(n = 1, p = .6, size = 20)
print(coinflips.mean())

stats.ttest_1samp(coinflips, .5)

0.4


Ttest_1sampResult(statistic=-0.8897565210026089, pvalue=0.3847242304314523)

In [66]:
coinflips = np.random.binomial(n = 1, p = .6, size = 20)
print(coinflips.mean())

stats.ttest_1samp(coinflips, .5)

0.7


Ttest_1sampResult(statistic=1.9023794624226835, pvalue=0.07239002577225488)

## Two Sample T-test

Comparing two groups or samples against each other rather than a single sample against a population. 



Ever thought about how long it takes to make a pancake? Have you ever compared the tooking time of a pancake on each eye of your stove? Is the cooking time different between the different eyes? Now, we can run an experiment and collect a sample of 1,000 pancakes on one eye and another 800 pancakes on the other eye. Assumed we used the same pan, batter, and technique on both eyes. Our average cooking times were 180 (5 std) and 178.5 (4.25 std) seconds repsectively. Now, we can tell those numbers are not identicial, but how confident are we that those numbers are practically the same? How do we know the slight difference isn't caused by some external randomness?

Yes, today's lesson will help you figure out how long to cook your pancakes (*theoretically*). Experimentation is up to you; otherwise, you have to accept my data as true. How are going to accomplish this? With probability, statistics, inference and maple syrup (optional). 

<img src="https://images.unsplash.com/photo-1541288097308-7b8e3f58c4c6?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=3300&q=80" width=400>



In [69]:
xbar1 = 180 # Sample Mean 1
s1 = 5 # Sample Std. Dev. 1
n1 = 1000

sample1 = np.random.normal(xbar1, s1, n1)

sample1[:5]

array([173.155424  , 183.1012723 , 181.62809258, 176.27824602,
       182.05190997])

In [71]:
xbar2 = 178.5 # Sample Mean 2
s2 = 4.25 # Sample Std. Dev. 2
n2 = 800

sample2 = np.random.normal(xbar2, s2, n2)

sample2[:5]

array([178.69566433, 177.27786253, 172.68707963, 178.16852352,
       183.12453756])

**Null:** Cooking times of the two burners is the same 

$\bar{x}_1 = \bar{x}_2$ (sample mean 1 is equal to sample mean 2)

**Alt:** Cooking times of the two burners is different

$\bar{x}_1 \neq \bar{x}_2$ (sample mean 1 is equal to sample mean 2)

**CI:** 95%

In [72]:
stats.ttest_ind(sample1, sample2)

Ttest_indResult(statistic=6.352994744904061, pvalue=2.6687441968298414e-10)

**Conclusion:** Based on a t-stat of 6.35 and a p-value of .0000000002668, I reject the null that the sample means are the same, and suggest the alt, that the two burners cook pancakes at different rates. 

Reject when P-value < (1 - Confidence Level)

## Lets Practice more 1-sample and 2-sample T-tests

Using the dataset that you will be using on your assignment

In [73]:
# Let's do it!

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-03-16 17:49:20--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-03-16 17:49:21 (278 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [83]:
names = ['party','handicapped-infants','water-project',
         'budget','physician-fee-freeze', 'el-salvador-aid',
         'religious-groups','anti-satellite-ban',
         'aid-to-contras','mx-missile','immigration',
         'synfuels', 'education', 'right-to-sue','crime','duty-free','south-africa']

df = pd.read_csv('house-votes-84.data', names = names)

print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [84]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

In [86]:
df = df.replace({'y':1, 'n':0, '?':np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
rep = df[df['party'] == 'republican']
dem = df[df['party'] == 'democrat']

## 1-sample T-Test

When you're doing a 1-sample t-test, you get to pick what your null hypothesis value is. You decide the conception of the population.

Republicans on the handicapped-infants bill

Null: Republicans are split on this issue 50-50%

Alt: Republicans are not split

CI: 95%

In [103]:
# I'll get nans for my t-test and p-value if there are NaNs in my sample
stats.ttest_1samp(rep['handicapped-infants'], .5) 

Ttest_1sampResult(statistic=nan, pvalue=nan)

In [104]:
# Dropping NaN's with 'omit' 
stats.ttest_1samp(rep['handicapped-infants'], .5, nan_policy = 'omit')

# Could alter this to test the entirity of congress of a specific subset by changing to the whole df

Ttest_1sampResult(statistic=-10.232833482397659, pvalue=2.572179359890009e-19)

**Conclusion:** Reject the null; republican support is different from a 50% split

In [105]:
rep['handicapped-infants'].mean()

0.18787878787878787

Can change the null value of .5 to something else to test other parameters

T-stat is negative, than the actual value is less than the null; not a bad heuristic

## 2-sample t-test (Null is always that the two samples are the same)

Null: Two samples are the same

Alt: Two samples are different 

CI: 95%

In [106]:
stats.ttest_ind(rep['handicapped-infants'], dem['handicapped-infants'], nan_policy = 'omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [107]:
rep['handicapped-infants'].mean()

0.18787878787878787

In [108]:
dem['handicapped-infants'].mean()

0.6046511627906976

## Student's T Test

>Assuming data come from a Normal distribution, the t test provides a way to test whether the sample mean (that is the mean calculated from the data) is a good estimate of the population mean. 

The derivation of the t-distribution was first published in 1908 by William Gosset while working for the Guinness Brewery in Dublin. Due to proprietary issues, he had to publish under a pseudonym, and so he used the name Student.

The t-distribution is essentially a distribution of means of normaly distributed data. When we use a t-statistic, we are  checking that a mean fails within a certain $\alpha$ probability of the mean of means.

# Resources

- https://homepage.divms.uiowa.edu/~mbognar/applets/t.html
- https://rpsychologist.com/d3/tdist/
- https://gallery.shinyapps.io/tdist/
- https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation_of_metabolic_rate_of_northern_fulmars
- https://www.khanacademy.org/math/ap-statistics/two-sample-inference/two-sample-t-test-means/v/two-sample-t-test-for-difference-of-means

Change so I can make pull request